dokka updated to 1.4.20

RingerJK opened the pull request:
### Description

Introducing Dokka v1.4+ 

A very first glance. PR shows how a new documentation would look like.
~~Blocked by `Dokka`'s alpha version. We're waiting for **stable** one.~~

Ref #3501 

Steps to run dokka it terminal: 
- $ `cd` to root of `mapbox-navigation-android`
- $ `make javadoc-dokka` (or `./gradlew dokkaHtmlMultiModule`)
- multi-module documentation location: `mapbox-navigation-android/build/kdoc`

### Screenshots or Gifs

#### Docs with `dokka` v0.10(previous version):
![Screen Shot 2020-11-10 at 19 52 21](https://user-images.githubusercontent.com/8738058/98704932-41ba7900-238e-11eb-9410-09b7ed96b574.png)

![Screen Shot 2020-11-10 at 19 54 08](https://user-images.githubusercontent.com/8738058/98705198-8219f700-238e-11eb-81b9-15bae6ba12af.png)

![Screen Shot 2020-11-10 at 19 54 59](https://user-images.githubusercontent.com/8738058/98705312-9f4ec580-238e-11eb-95bc-6e7c40275675.png)


#### Docs with `dokka` v1.4+ 
![Screen Shot 2020-11-10 at 19 53 18](https://user-images.githubusercontent.com/8738058/98705096-631b6500-238e-11eb-87f6-fc66028efd00.png)

![Screen Shot 2020-11-10 at 19 55 09](https://user-images.githubusercontent.com/8738058/98705331-a675d380-238e-11eb-9806-8e11981bd3dc.png)

![Screen Shot 2020-11-10 at 19 55 26](https://user-images.githubusercontent.com/8738058/98705359-af66a500-238e-11eb-8368-f0ebdbecdfff.png)

<!-- Include media files to provide additional context. It's REALLY useful for UI related PRs (e.g. ![screenshot gif](link)) -->


<!--
---------- CHECKLIST ----------
1. Add related labels (`bug`, `feature`, `new API(s)`, `SEMVER-MAJOR`, `needs-backporting`, etc.).
2. Update progress status on the project board.
3. Request a review from the team, if not a draft.
4. Add targeted milestone, when applicable.
5. Create ticket tracking addition of public documentation pages entry, when applicable.
-->

LukasPaczos commented on libnavigation-base/build.gradle:
```diff
@@ -4,15 +4,17 @@ apply plugin: 'org.jetbrains.dokka'
-            suppress = true
+dokkaHtml {
+    outputDirectory.set(new File(buildDir.path, "kdoc"))
+    moduleName.set("libnavigation-base")
```
Let's use more public-facing names, like `Base`. `libnavigation-base` is just an internal name for a module. Same everywhere below.

RingerJK replied:
sounds good!

LukasPaczos replied:
Just a reminder about this :)

LukasPaczos reviewed (commented):
Why did we lose the mapbox branding on top of the page? Can we restore it?

LukasPaczos commented:
I also wouldn't consider us blocked because of the alpha version of dokka, if it works reliably and unblocks our use-case, let's migrate. It's not a critical element and we can monitor the behavior as we go and upgrade to stable.

RingerJK commented:
> I also wouldn't consider us blocked because of the alpha version of dokka, if it works reliably and unblocks our use-case, let's migrate. It's not a critical element and we can monitor the behavior as we go and upgrade to stable.

I'm worrying about some bugs that we cannot check with alpha, you know. To check if everything is ok, we should go page by page manually 🚶

RingerJK commented:
> Why did we lose the mapbox branding on top of the page? Can we restore it?

Dokka doesn't support it(yet), we have to do it on our side. Believe we need support from docs-team. 
**cc** @danswick 

btw we could replace default top logo, [link](https://kotlin.github.io/dokka/1.4.10.2/user_guide/base-specific/frontend/#modifying-assets)

Guardiola31337 commented:
> Dokka doesn't support it(yet), we have to do it on our side. Believe we need support from docs-team.

We have Mapbox styles already. Can't we inject them somehow? Also noting that there's a script we can run to format the Dokka pages.

cc @danswick

danswick commented:
Woah, this looks really promising! Haven't had a chance to look closely yet, but looking forward to. Since this new format looks a lot more polished, we might only need to do a few tweaks. Happy to look into that more as well.

LukasPaczos commented:
> we might only need to do a few tweaks

What tweaks do you have in mind? Let's add them!

Guardiola31337 commented:
> What tweaks do you have in mind? Let's add them!

+1000

cc @danswick

danswick commented:
I think we'll at least want to add the Mapbox logo and maybe some CSS to make things look a little more like the rest of our sites. If anyone would be willing to pair, maybe we could walk through how things are set up together and I could open a PR here.

danswick commented:
@RingerJK I added a small script in https://github.com/mapbox/mapbox-navigation-android/pull/3757/commits/f08d8a8943d778639dbac4bd420fbec43b88f012 and https://github.com/mapbox/mapbox-navigation-android/pull/3757/commits/d3b3635e0fd7ae12a33cee1fb5f6acabfae03cf8 that finds and replaces all of the `logo-styles.css` files during `make javadoc-dokka`. 

Ideally, we would configure the custom logo location in Dokka's base plugin, but I just don't know enough about Dokka or Kotlin or Gradle (you get the idea 🙃) to figure out how to do it. I tried working off of the changes I found in https://github.com/Kotlin/dokka/pull/1488, but was not able to get the build to work. A simple script at least does the trick for now and I was able to see the logo on all of the module pages.

Let me know what you think!

RingerJK commented:
@danswick thanks for addressing comments 😃 

> A simple script at least does the trick for now and I was able to see the logo on all of the module pages.

According dokka' wiki, script is an only one possible way to replace logo properly 😜 

Seems we're missing mapbox' styles and pages' headers only 👍

danswick commented:
@RingerJK In my opinion, the dokka docs are fine in their current state and we don't need to do any additional styling. The docs team is planning to work on generated docs styling and some related work in Q1 2020. I think we should just go with this configuration for now. 

> According dokka' wiki, script is an only one possible way to replace logo properly

Custom styling with Dokka is very confusing, but I believe you can load in any CSS you want (including logo overrides) by configuring the Dokka base plugin. I have no idea how to do that, though 🤷

RingerJK commented:
@danswick thanks, get it! where can we track work of docs styling? 

Also need a support from docs team https://github.com/mapbox/mapbox-navigation-android/issues/3501#issuecomment-729759591 : redesign navigation api-refs page(it should separate docs on ui and non-ui starting from 1.0)

LukasPaczos commented on scripts/dokka-validate.sh:
```diff
@@ -2,22 +2,23 @@
 ./gradlew clean
 echo "Dokka is validating ${MODULE}"
-./gradlew "${MODULE}dokka" | grep -i "$SEARCH_REGEX" | grep -v "$SEARCH_REGEX_EXCEPT" && { echo 'kdoc validation failed'; ${EXIT_TYPE}; }
+./gradlew $GRADLE_TASK | grep -i "$SEARCH_REGEX" #| grep -v "$SEARCH_REGEX_EXCEPT" && { echo 'kdoc validation failed'; ${EXIT_TYPE}; }
```
Since we are commenting out the `exit 1` call, this script will never fail, isn't that right?

RingerJK replied:
I was testing new approach. revert it 👍

RingerJK commented:
@LukasPaczos @Guardiola31337 ready for a new round

LukasPaczos commented on libnavigator/build.gradle:
```diff
@@ -4,15 +4,17 @@ apply plugin: 'org.jetbrains.dokka'
-        reportUndocumented = true
+dokkaHtml {
+    outputDirectory.set(new File(buildDir.path, "kdoc"))
+    moduleName.set("LibNavigator")
```
All of the classes in this module are internal (or in an internal package), can we remove this module from being published to the website?

RingerJK replied:
@LukasPaczos The same applicable for `Notification` and `Router` modules. Problem here that we don't have a mechanism to track if public api in modules appear without dokka. So here like a trade off: - we publish a whole module and keep tracking it - or remove it from dokka set and doesn't have automative mechanism.

LukasPaczos replied:
Can we not specify which modules should be published as part of dokka's multimodule task?

>  Problem here that we don't have a mechanism to track if public api in modules without dokka

Regardless, since all classes are internal or in internal packages in those modules, dokka task does not iterate over them anyway, right? So we should be fine to remove the task without actually losing any value.

RingerJK replied:
> Regardless, since all classes are internal or in internal packages in those modules, dokka task does not iterate over them anyway, right? So we should be fine to remove the task without actually losing any value.

I'm agree with it. For customers perspective it shouldn't shared at all 👍 


> Can we not specify which modules should be published as part of dokka's multimodule task?
actually we're not specify it outside of a module. Each module that use dokka plugin automatically consumed by dokka-multimodule task.

RingerJK replied:
not possible just remove modules that shouldn't be a part of api-docs, `publish` configuration plugin depends on dokka 
https://github.com/mapbox/mapbox-navigation-android/blob/5b325bb67ded6b74f8c690f0318713e455a589a2/gradle/publish.gradle#L15
let's merge as it is. I'll add it to ticket to split work

LukasPaczos commented on scripts/dokka-validate.sh:
```diff
@@ -2,22 +2,33 @@
 
-SEARCH_REGEX='No documentation for com.mapbox.navigation'
-SEARCH_REGEX_EXCEPT='\.Companion ' # skip `Companion object` headers
+fileItemString=$(cat  scripts/undocumented_classes.txt |tr "\n" " ")
```
Do we allow for the public APIs in those classes to remain undocumented?

If that's the case, could you cut a ticket to fill up the gaps?

RingerJK replied:
> Do we allow for the public APIs in those classes to remain undocumented?

We didn't have a chance to track java classes before dokka 1.4+. 

I was going to cut a ticket as this pr is landed 👍
